scipy/optimize/jonkervolgenant/jonkervolgenant.cpp

```diff
@@ -0,0 +1,194 @@
+                std::vector<double>& shortestPathCosts, int i,
+                std::vector<bool>& SR, std::vector<bool>& SC, double* p_minVal)
+{
+    double minVal = 0;
+
+    int num_remaining = nc;
+    std::vector<int> remaining(nc);
+    // std::iota(remaining.begin(), remaining.end(), 0); //needs C++11
+    for (int it = 0; it < nc; it++) {
+        remaining[it] = it;
+    }
+
+    std::fill(SR.begin(), SR.end(), false);
+    std::fill(SC.begin(), SC.end(), false);
+    std::fill(shortestPathCosts.begin(), shortestPathCosts.end(), INFINITY);
+
+    // find shortest augmenting path
+    int sink = -1;
+    while (sink == -1) {
+
+        int index = -1;
+        double lowest = INFINITY;
+
+        for (int it = 0; it < num_remaining; it++) {
+            int j = remaining[it];
+
+            double r = minVal + cost[i * nc + j] - u[i] - v[j];
+            if (r < shortestPathCosts[j]) {
+                path[j] = i;
+                shortestPathCosts[j] = r;
+            }
+
+            if (shortestPathCosts[j] < lowest ||
+                (shortestPathCosts[j] == lowest && row4col[j] == -1)) {
+                lowest = shortestPathCosts[j];
+                index = it;
+            }
+        }
+
+        minVal = lowest;
+        int j = remaining[index];
+        if (minVal == INFINITY) { // infeasible cost matrix
+            return -1;
+        }
+
+        if (row4col[j] == -1) {
+            sink = j;
+        } else {
+            i = row4col[j];
+        }
+
+        SR[i] = true;
+        SC[j] = true;
+        remaining[index] = remaining[--num_remaining];
+        remaining.resize(num_remaining);
+    }
+
+    *p_minVal = minVal;
+    return sink;
+}
+
+static int
+solve(int nr, int nc, double* input_cost, int64_t* output_col4row)
+{
+
+    // build a non-negative cost matrix
+    std::vector<double> cost(nr * nc);
+    double minval = *std::min_element(input_cost, input_cost + nr * nc);
+    for (int i = 0; i < nr * nc; i++) {
+        cost[i] = input_cost[i] - minval;
+    }
+
+    // initialize variables
+    std::vector<double> u(nr, 0);
+    std::vector<double> v(nc, 0);
+    std::vector<double> shortestPathCosts(nc);
+    std::vector<int> path(nc, -1);
+    std::vector<int> col4row(nr, -1);
+    std::vector<int> row4col(nc, -1);
+    std::vector<bool> SR(nr);
+    std::vector<bool> SC(nc);
+
+    // iteratively build the solution
+    for (int curRow = 0; curRow < nr; curRow++) {
+
+        double minVal;
+        int sink = augmenting_path(nc, cost, u, v, path, row4col,
+                                   shortestPathCosts, curRow, SR, SC, &minVal);
+        if (sink < 0) {
+            return -1;
+        }
+
+        // update dual variables
+        u[curRow] += minVal;
+        for (int i = 0; i < nr; i++) {
+            if (SR[i] && i != curRow) {
+                u[i] += minVal - shortestPathCosts[col4row[i]];
+            }
+        }
+
+        for (int i = 0; i < nc; i++) {
```

Comment:
Probably a matter of taste, but I would find this ever so slightly easier to read, if columns were consistently indexed by `j` here.

Reply:
OK.  The pseudocode uses this indexing, so I have made the change.